### ISSUE TITLE
UndefinedDocblockClass error for @psalm-type annotation when class with public or protected property is extended

### ISSUE THREAD

**Issue by theodorejb** [vimeo/psalm#8645]
This is a very bizarre bug. I wasn't able to reproduce it on https://psalm.dev/ since it only occurs when the classes are in separate files.

> ERROR: UndefinedDocblockClass - src/A.php:19:9 - Docblock-defined class, interface or enum named Test\Foo does not exist

> ERROR: InvalidArgument - src/A.php:20:23 - Argument 1 of Test\A::useFoo expects array{x: string, y: string}, but Test\Foo provided

The errors occur both in Psalm 4.x (I tested 4.29 and 4.19), as well as the latest master branch for Psalm 5 (commit 7c83878ae00b21959ba49d084cc1ddbb66e0d730).

Minimal reproduction:

`A.php`:

```php
<?php

namespace Test;

/**
 * @psalm-type Foo = array{x: string, y: string}
 */
class A
{
    // Error only occurs if this property is public or protected, and does not have a default value.
    // The property type doesn't matter.
    protected int $prop;

    public function __construct()
    {
        $this->prop = 1;

        /** @var Foo $foo */
        $foo = array_replace_recursive(['y' => ''], ['x' => '']);
        $this->useFoo($foo);
    }

    /**
     * @param Foo $foo
     */
    public function useFoo(array $foo): void
    {
        var_export($foo);
    }
}
```

`B.php`:

```php
<?php

namespace Test;

// error only occurs if this class is in a separate file and extends A
class B extends A
{
}
```

I also set up a repository to reproduce the issue in case that's helpful: https://github.com/theodorejb/psalm-type-bug

**Comment by danog** [comment]
Side note, the `@var` annotation is pointless here, you can easily pass this array to functions that take a `Foo` even without the `@var`, in psalm type aliases can be equal to the underlying type.

**Comment by theodorejb** [comment]
@danog When it comes to my minimal reproduction, you are right. But in my actual project the annotation is necessary since it uses `array_replace_recursive` similar to this: https://psalm.dev/r/915dfb04cf.

**Comment by theodorejb** [comment]
I did some more digging, and found that the error was introduced by https://github.com/vimeo/psalm/commit/ff00255c194cebc84145e86e2376111d39f2801c (PR #5628), first released in Psalm 4.7.1.

**Comment by orklah** [comment]
Can you post the result of the psalm analysis with --debug-by-line please?

**Comment by theodorejb** [comment]
@orklah Here are the results with `--debug-by-line`: 
[debug.txt](https://github.com/vimeo/psalm/files/9929694/debug.txt)

**Comment by orklah** [comment]
Can you check if adding 
```
/**
 * @psalm-import-type Foo from A
 */
```

In B solves the issue?

**Comment by theodorejb** [comment]
@orklah Yes, strangely the error does disappear if I add that type import.

**Comment by orklah** [comment]
That's weird Psalm needs that, not sure if that was how it was designed initially. I'll tag this as enhancement

**Comment by theodorejb** [comment]
I don't think it makes sense for sub-classes to have to import a type, if the subclass doesn't use it anywhere. It's especially confusing because the error only occurs when the subclass is in a different file.
